qrexec/server.py

```diff
@@ -0,0 +1,118 @@
+        self._socket_path = socket_path
+        self._socket_activated = socket_activated
+
+    async def run(self):
+        server = await self.start()
+        async with server:
+            await server.serve_forever()
+
+    async def start(self):
+        if self._socket_activated:
+            fds = listen_fds()
+            if fds:
+                assert len(fds) == 1, 'too many listen_fds: {}'.format(
+                    listen_fds)
+                sock = socket.socket(fileno=fds[0])
+                return await asyncio.start_unix_server(self._client_connected,
+                                                       sock=sock)
+
+        if os.path.exists(self._socket_path):
+            os.unlink(self._socket_path)
+        return await asyncio.start_unix_server(self._client_connected,
+                                               path=self._socket_path)
+
+    async def _client_connected(self, reader, writer):
+        try:
+            data = await reader.read()
+            data = data.decode('ascii')
+            assert '\0' in data, data
+
+            header, json_data = data.split('\0', 1)
+            service, source_domain = header.split(' ')
+            params = json.loads(json_data)
+
+            response = await self.handle_request(params, service, source_domain)
+
+            writer.write(response.encode('ascii'))
+            await writer.drain()
+        finally:
+            writer.close()
+            await writer.wait_closed()
+
+    async def handle_request(self, params, service, source_domain):
+        raise NotImplementedError()
+
+
+def call_socket_service(
+        remote_domain, service, source_domain, params,
+        rpc_path=RPC_PATH):
+    if remote_domain == 'dom0':
+        return call_socket_service_local(
+            service, source_domain, params, rpc_path)
+    return call_socket_service_remote(
+        remote_domain, service, source_domain, params)
+
+
+async def call_socket_service_local(service, source_domain, params,
+                                    rpc_path=RPC_PATH):
+    path = os.path.join(rpc_path, service)
+    reader, writer = await asyncio.open_unix_connection(path)
+    writer.write('{} {}\0'.format(service, source_domain).encode('ascii'))
+    writer.write(json.dumps(params).encode('ascii'))
+    writer.write_eof()
+    await writer.drain()
+    response = await reader.read()
+    return response.decode('ascii')
+
+
+async def call_socket_service_remote(remote_domain, service, source_domain,
+                                     params):
+    qrexec_opts = ['-d', remote_domain]
+    cmd = 'DEFAULT:QUBESRPC {} {}'.format(service, source_domain)
+    command = [QREXEC_CLIENT] + qrexec_opts + [cmd]
+    process = await asyncio.create_subprocess_exec(
+        *command,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE)
+    stdin, _stderr = await process.communicate(
+        json.dumps(params).encode('ascii'))
+    return stdin.decode('ascii')
```

Comment:
There is already `qrexec.client_dom0` and `qrexec.client_vm` (and a code in __init__.py to choose one based on where it is called). They are different (not async def, no json encoding, no direct socket support for local calls), so not exactly duplication, but IMO worth keeping them in the same place, and document differences, to avoid confusion. Also, distinguishing local from remote based on target being dom0 works only in dom0...

Reply:
I extracted the generic part to `qrexec.client_dom0` and `qrexec.client_vm`, and documented the limitations of current module.